### DISCUSSION TITLE
IsAssignableTo on Enums

### DISCUSSION THREAD

**Discussion by AnakinRaW** [dotnet/runtime#60547] [Q&A] [answered]
While reading ECMA 335 I noticed that in I.8.5.2
```
The CTS supports an enum [...] Instances of an enum, however, shall be assignable-to the underlying type, and
vice versa. That is, no cast (see §I.8.3.3) or coercion (see §I.8.3.2) is required to convert from the
enum to the underlying type [...]
```
However if i run 
`typeof(IntEnum).IsAssignableTo(typeof(int))` it gives me `false` as return. Also the CastingHelper::CanCastTo in src\coreclr\tools\common\typesystem return `false`. 

Anybody can help me what I'm missing here.
Cheers

**Comment by sakno** [comment, marked as answer]
Because of strong typing: enum inherits from `Enum` class while `Int32` inherits from `ValueType`. Moreover, some methods behave differently. For instance,
```csharp
Console.WriteLine(EnvironmentVariableTarget.User.ToString());
Console.WriteLine(1.ToString());
```

The binary representations of both values are same, that's why cast or coercion is not needed. That is, at IL and machine code level there is no differences between them. However, event C# requires the cast because the cast changes the type. The type in OOP is not only the data encapsulated by the instance but also the behavior. That's why you need that cast. Just like `Int32` and `UInt32`. Both contain 4-bytes long data, but behave differently.

`IsAssignableTo` checks whether the cast doesn't change the behavior. The only way to do this is to check _upcast_ - casting from derived type to one of its base types.

**Reply by AnakinRaW** [reply]
Had to read that a couple of times, but it makes sense. Thanks :)!
